### PR TITLE
fuse -> ROS 2 fuse_graphs : Port fuse_graphs

### PR DESCRIPTION
**Pull request by methylDragon**
See: https://github.com/locusrobotics/fuse/issues/276

# Description

This PR ports the entirety of fuse_graphs, including:
- CMakeLists and packages
- Source code
- Tests

Other stuff like docs are in other PRs.
Some method signatures were altered in fuse_core to support the PR.

# Notes
- This PR is branched off the branch from https://github.com/locusrobotics/fuse/pull/288
	- i.e. https://github.com/locusrobotics/fuse/pull/288 should be merged FIRST
	- (otherwise merging this will merge 288 (and 287 and 286) as well)

Pinging @svwilliams for visibility.

### CODE REVIEW

File: fuse_core/include/fuse_core/graph.h [hunk 1]
```diff
@@ -381,7 +381,8 @@ class Graph
    */
   virtual ceres::Solver::Summary optimizeFor(
     const rclcpp::Duration& max_optimization_time,
-    const ceres::Solver::Options& options = ceres::Solver::Options()) = 0;
+    const ceres::Solver::Options& options = ceres::Solver::Options(),
+    rclcpp::Clock clock = rclcpp::Clock(RCL_STEADY_TIME)) = 0;  // NOTE(CH3): We need to copy because clock.now() is non-const...
```
Comment (sloretz) [inline fuse_core/include/fuse_core/graph.h]: Opened PR to make `rclcpp::Clock::now()` const: https://github.com/ros2/rclcpp/pull/2050

File: fuse_graphs/CMakeLists.txt [hunk 2]
```diff
@@ -7,136 +7,93 @@ set(build_depends
...
+include(boost-extras.cmake)
 
 ###########
 ## Build ##
 ###########
-add_compile_options(-Wall -Werror)
 
 ## fuse_graphs library
-add_library(${PROJECT_NAME}
+add_library(${PROJECT_NAME} SHARED
```
Comment (sloretz) [inline fuse_graphs/CMakeLists.txt]: Same comment about leaving off `SHARED` and using `ament_cmake_ros` https://github.com/locusrobotics/fuse/pull/287#discussion_r1028631887

File: fuse_graphs/CMakeLists.txt [hunk 3]
```diff
@@ -7,136 +7,93 @@ set(build_depends
...
+  RUNTIME DESTINATION bin
+)
+
+install(DIRECTORY include/
+  DESTINATION include/${PROJECT_NAME}
+)
+
+install(
+  FILES fuse_plugins.xml
+  DESTINATION share/${PROJECT_NAME}
```
Comment (sloretz) [inline fuse_graphs/CMakeLists.txt]: Same comment about using `pluginlib_export_plugin_description_file()` https://github.com/locusrobotics/fuse/pull/287#discussion_r1028630684

File: fuse_graphs/CMakeLists.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.0.2)
+cmake_minimum_required(VERSION 3.16)
 project(fuse_graphs)
 
 set(build_depends
```
Comment (sloretz) [inline fuse_graphs/CMakeLists.txt]: nit, unused?

File: fuse_graphs/package.xml [hunk 5]
```diff
@@ -13,16 +13,25 @@
...
 
-  <buildtool_depend>catkin</buildtool_depend>
-  <depend>libceres-dev</depend>
-  <depend>fuse_core</depend>
-  <depend>pluginlib</depend>
-  <depend>roscpp</depend>
-  <test_depend condition="$ROS_DISTRO >= melodic">benchmark</test_depend>
-  <test_depend>roslint</test_depend>
-  <test_depend>rostest</test_depend>
+  <buildtool_depend>ament_cmake</buildtool_depend>
```
Comment (sloretz) [inline fuse_graphs/package.xml]: nit, ament_cmake_ros

File: fuse_graphs/benchmark/example_variable.h [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FUSE_GRAPHS_TEST_EXAMPLE_VARIABLE_H  // NOLINT{build/header_guard}
```
Comment (sloretz) [inline fuse_graphs/benchmark/example_variable.h]: Accidental copy of `fuse_graphs/tests/example_variable.h`?

Comment (methylDragon) [reply fuse_graphs/benchmark/example_variable.h]: How did that happen :thinking: 
I hope it wasn't my script doing it :/

**Comment by methylDragon** [comment]
@ros-pull-request-builder retest this please!

File: fuse_graphs/benchmark/example_variable.h (hunk 6 above)
Comment (sloretz) [reply fuse_graphs/benchmark/example_variable.h]: Ah looks like it was being relied on by the benchmark. Adding the `test` directory as an include path should fix it


```
14:36:52 /tmp/ws/src/fuse/fuse_graphs/benchmark/benchmark_create_problem.cpp:39:10: fatal error: example_variable.h: No such file or directory
14:36:52    39 | #include "example_variable.h"
```

```CMake
target_include_directories(benchmark_create_problem PRIVATE "test/")
```

Comment (methylDragon) [reply fuse_graphs/benchmark/example_variable.h]: Ah, you're right. I duplicated it because of that, but making it rely on test's version works too.

**Comment by methylDragon** [comment]
@ros-pull-request-builder retest this please!

[Build Status](https://build.ros2.org/job/Rpr__fuse__ubuntu_jammy_amd64/23/)

File: fuse_graphs/benchmark/example_variable.h (hunk 6 above)
Comment (methylDragon) [reply fuse_graphs/benchmark/example_variable.h]: https://github.com/locusrobotics/fuse/compare/b1c87003ab7a2a16bd928502d71cc3e58b9382bd..38006f30cf78a25d0a07ee8a93ab733a788a6594

**Review by sloretz — APPROVED** [review]
LGTM with RPr job where only linters for `fuse_graphs` are failing

**Comment by sloretz** [comment]
Ah I see, https://build.ros2.org/job/Rpr__fuse__ubuntu_jammy_amd64/23/#showFailuresLink is for this PR (I see 38006f3 being checked out in the console text). Looks ready to merge to me!

**Pull request merged**